ocrd/ocrd/processor/base.py

```diff
@@ -147,33 +147,34 @@ def add_metadata(self, pcgts):
                                          value=OCRD_VERSION)])
                     ]))
 
-    def resolve_resource(self, parameter_name, val):
+    def resolve_resource(self, val):
         """
         Resolve a resource name to an absolute file path with the algorithm in
         https://ocr-d.de/en/spec/ocrd_tool#file-parameters
 
         Args:
-            parameter_name (string): name of parameter to resolve resource for
             val (string): resource value to resolve
         """
         executable = self.ocrd_tool['executable']
-        try:
-            param = self.ocrd_tool['parameter'][parameter_name]
-        except KeyError:
-            raise ValueError("Parameter '%s' not defined in ocrd-tool.json" % parameter_name)
-        if not param['mimetype']:
-            raise ValueError("Parameter '%s' is not a file parameter (has no 'mimetype' field)" %
-                             parameter_name)
-        if val.startswith('http:') or val.startswith('https:'):
-            return OcrdResourceManager().download(executable, val)
-        ret = next([cand for cand in list_resource_candidates(executable, val) if exists(cand)])
+        if exists(val):
+            return val
+        ret = [cand for cand in list_resource_candidates(executable, val) if exists(cand)]
         if ret:
-            return ret
-        bundled_fpath = resource_filename(__name__, val)
```

Comment:
Why did you remove the `pkg_resources.resource_filename` (Python fall-back) mechanism entirely? Isn't this even required by the spec now?

(I know it's harder to get these under the centralized `ocrd resmgr list-installed`. But if `Processor.list_all_resources()` would append the files from pkg_resources, then the former could delegate to these via CLI calls.)

Reply:
> Why did you remove the pkg_resources.resource_filename (Python fall-back) mechanism again entirely? Isn't this even required by the spec now?

To be honest, I was trying to get this right for ocrd_typegroups_classifier but I found it more intuitive to offer the bundled resource as a regular resource in `resources_list.yml` (since it is already in GitHub and hence HTTP-downloadable) and let ocrd_typegroups_classifier handle the case where no nerwork parameter was given itself. It's a bit of a waste of space but makes the code for `list_all_resources` and `resmgr` simpler.

It would be good to have an example project that makes use of bundled resources to test this with and if we don't, it would be better to change the spec TBH. 

BTW, I wanted to also list https://github.com/ASVLeipzig/cor-asv-ann-models but gave up on getting a "lightweight" solution working to download from LFS.

Reply:
> > Why did you remove the pkg_resources.resource_filename (Python fall-back) mechanism again entirely? Isn't this even required by the spec now?
> 
> To be honest, I was trying to get this right for ocrd_typegroups_classifier but I found it more intuitive to offer the bundled resource as a regular resource in `resources_list.yml` (since it is already in GitHub and hence HTTP-downloadable) and let ocrd_typegroups_classifier handle the case where no nerwork parameter was given itself. It's a bit of a waste of space but makes the code for `list_all_resources` and `resmgr` simpler.

I see. Well, `ocrd_typegroups_classifier` or [document-image-binarization](https://github.com/ajgallego/document-image-binarization) are examples of packages which distribute their model files (not sure about binary vs sdist installation though), so it _should_ already be available locally. But I agree that the waste of bandwidth/space is probably not a big deal in general because if the models are too large, devs will not include them in the package (or even the same repo) anyway.

For modules with very large files like `ocrd_anybaseocr` – not all of which might be needed – I guess it makes no sense to expect them to be distributed along with the base pkg. (Although the PyPI server could still help distribute dependent packages like `ocrd-anybaseocr-models-dewarping`, `ocrd-anybaseocr-models-segmentation` etc.)

But then there's modules like ocrd_wrap: I want to put many small preset files there in such a way that they can be referenced on the CLI  (`-p param_im6convert-denoise-raw.json -P level-of-operation region`). But if that's (going to be) possible by just doing `ocrd resmgr download ocrd-preprocess-image '*'` (which I could even delegate to in my `make install` rule), then fine. (I might even register resources for external tools there.)

> It would be good to have an example project that makes use of bundled resources to test this with and if we don't, it would be better to change the spec TBH.

Yes, agreed: If we don't actually need package resources as a mechanism, we shouldn't have it in the spec. Wrapping an existing pkg resource to an OCR-D file resource spec should be easy.

But that's actually something I still don't see here: we have a _centralized_ `resource_list.yml` now, but not a way for modules to feed their contributions...

> BTW, I wanted to also list https://github.com/ASVLeipzig/cor-asv-ann-models but gave up on getting a "lightweight" solution working to download from LFS.

I'll have a look myself later-on. Github LFS always gave us problems in the past already (we usually exceed the monthly quota without even knowing why). There's still the Gitlab backup, but...

Reply:
> But that's actually something I still don't see here: we have a centralized resource_list.yml now, but not a way for modules to feed their contributions...

There is the option of a user-defined list in addition to the centralized list. Another option would be to extend the ocrd-tool.json schema so that project authors can themselve define the resources and we introspect that at runtime?

Reply:
> Another option would be to extend the ocrd-tool.json schema so that project authors can themselve define the resources and we introspect that at runtime?

Excellent idea! So ocrd_all installation rules could aggregate ocrd's `resource_list.yml` or `~/.config/ocrd/resource.yml` from that (enabling `make install-models` after `make all`). But also, each individual module makefile could get a local `install-models` rule...

Reply:
Since integrating this into ocrd-tool.json requires changes to the spec and we should probably adapt the metadata scheme to be interoperable with [@mittagessen's schema](https://github.com/mittagessen/kraken/blob/master/kraken/metadata.schema.json) I will do that in a separate set of PRs.

Reply:
So IIUC you want to merge the centralized resource list first, and adapt ocrd_all, and then work towards decentralized, aggregated resource definitions in tool.json?

But the Kraken model schema does not even have a URL attribute! Or are you planning for some schema extension? And isn't it specific to OCR models (`script`, `graphemes`)? Also, `accuracy` by itself does not make sense to me. At least, some of these features should not be `required`...

Reply:
accuracy: Yes, the method / metric often is a parameter itself, but more importantly, different models will usually also have different test data...

Kraken, Okralact, Calamari config, ... so many options. I don't think we should even try to use some interoperable / unified scheme.